crates/nu-command/src/formats/from/json.rs

```diff
@@ -56,49 +68,75 @@ impl Command for FromJson {
         input: PipelineData,
     ) -> Result<PipelineData, ShellError> {
         let span = call.head;
-        let (string_input, span, metadata) = input.collect_string_strict(span)?;
-
-        if string_input.is_empty() {
-            return Ok(Value::nothing(span).into_pipeline_data());
-        }
 
         let strict = call.has_flag(engine_state, stack, "strict")?;
 
         // TODO: turn this into a structured underline of the nu_json error
         if call.has_flag(engine_state, stack, "objects")? {
-            let lines = string_input.lines().filter(|line| !line.trim().is_empty());
-
-            let converted_lines: Vec<_> = if strict {
-                lines
-                    .map(|line| {
-                        convert_string_to_value_strict(line, span)
-                            .unwrap_or_else(|err| Value::error(err, span))
-                    })
-                    .collect()
-            } else {
-                lines
-                    .map(|line| {
-                        convert_string_to_value(line, span)
-                            .unwrap_or_else(|err| Value::error(err, span))
-                    })
-                    .collect()
-            };
-
-            Ok(converted_lines.into_pipeline_data_with_metadata(
-                span,
-                engine_state.ctrlc.clone(),
-                metadata,
-            ))
-        } else if strict {
-            Ok(convert_string_to_value_strict(&string_input, span)?
-                .into_pipeline_data_with_metadata(metadata))
+            // Return a stream of JSON values, one for each non-empty line
+            match input {
+                PipelineData::Value(Value::String { val, .. }, metadata) => {
+                    Ok(PipelineData::ListStream(
+                        read_json_lines(Cursor::new(val.into_bytes()), span, strict),
```

Comment:
Also, should we add a `ctrlc` handler here? (It might need to be passed to `read_json_lines`.)

Reply:
Oh, yes, actually.

I wonder if there's a way to make that less of a pitfall... ideally I'd love to have some kind of type-protected way to ensure that something that should have an interrupt.

I've thought about a marker trait like `trait Interruptible`  that can be applied to readers that have ctrlc, and adapters of readers that have it, and then you could just wrap a reader with `WithInterrupt<R>` that passes through `Read` and `BufRead` but checks ctrlc for you.

Reply:
> ideally I'd love to have some kind of type-protected way to ensure that something that should have an interrupt.

Agreed, I was thinking of maybe creating an `Interrupt` type and requiring all stream to take one at creation (i.e. not `Option`al). Then, combine this with some methods on `Bytestream` similar to `ListStream::modify` and maybe adapters to convert between the two streams, and this should hopefully help solve the issue. Although, I;m not sure if this covers every case.

Anyways... thanks for the change/commit :smile: